lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,253 @@
+from opaque_keys.edx.locator import BlockUsageLocator
+from xmodule_django.models import CourseKeyField, UsageKeyField
+
+from hashlib import sha256
+import json
+
+
+class VisibleBlocks(models.Model):
+    """
+    A django model used to track the state of a set of visible blocks under a given subsection at the time they are
+    used for grade calculation.
+
+    This state is represented using an array of serialized BlockRecords, stored in the blocks_json field. A
+    hash of this json array is used for lookup purposes.
+    """
+    _blocks_json = models.TextField(db_column="blocks_json")
+    hashed = models.CharField(max_length=32, primary_key=True)
+
+    def __unicode__(self):
+        """
+        String representation of this model.
+        """
+        return self.hashed
+
+    @classmethod
+    def create(cls, blocks):
+        """
+        Creates a new VisibleBlocks model. Argument 'blocks' should be an iterable collection of BlockRecords.
+        """
+        # Start by sorting the blocks to ensure equality regardless of how the blocks are ordered
+        sorted_blocks = sorted(
+            blocks,
+            key=lambda block: '{}{}{}'.format(block.block_key, block.max_score, block.weight)
+        )
+        _blocks_json = json.dumps(
+            [
+                block.to_dict()
+                for block in sorted_blocks
+            ]
+        )
+        hashed = sha256(_blocks_json).hexdigest()
+        model, _ = cls.objects.get_or_create(hashed=hashed, defaults={'_blocks_json': _blocks_json})
+        return model
+
+    @property
+    def blocks(self):
+        """
+        Returns the blocks_json data stored on this model as an list of BlockRecords. If for some reason block_json
+        is not parsable, a json error will bubble up.
+        """
+        block_dicts = json.loads(self._blocks_json)
+        return [BlockRecord(data['weight'], data['max_score'], data['block_key']) for data in block_dicts]
+
+    @blocks.setter
+    def blocks(self, value):
+        """
+        Not implemented, as VisibleBlocks instances are intended to be write-once and not change after creation.
+        """
+        raise NotImplementedError(
+            "Property 'blocks' cannot be modified on an existing VisibleBlocks model. Create a new instance."
+        )
+
+
+class BlockRecord(object):
+    """
+    An object encapsulating all relevant information for representing a block at the time it was used for grade
+    calculation.
+    """
+
+    def __init__(self, weight, max_score, locator):
+        """
+        Creates a BlockRecord.
+
+        Params:
+            weight (float)
+            max_score (float)
+            locator (BlockUsageLocator or string)
+        """
+        self.weight = weight
+        self.max_score = max_score
+        # pylint: disable=protected-access
+        self.block_key = locator._to_string() if isinstance(locator, BlockUsageLocator) else locator
```

Comment:
Please add a comment why this line is necessary. If we always want a string, can't we say, `self.block_key = unicode(locator)`? But why not just require that people use locators?